include/crocoddyl/core/costs/residual.hxx

```diff
@@ -0,0 +1,87 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2021, University of Edinburgh
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#include "crocoddyl/core/utils/exception.hpp"
+
+namespace crocoddyl {
+template <typename Scalar>
+CostModelResidualTpl<Scalar>::CostModelResidualTpl(boost::shared_ptr<typename Base::StateAbstract> state,
+                                                   boost::shared_ptr<ResidualModelAbstract> residual,
+                                                   boost::shared_ptr<ActivationModelAbstract> activation)
```

Comment:
I think it is better to keep the `CostModelAbstract` class, as 1) it allows us to develop any cost function without the `residual` assumption (i.e., Newton-Gauss approximation), and 2) it does not break the API.

On the other hand, we could define `calc` and `calcDiff` as nonpure virtual functions (as an action to handle the above-mentioned points), but then the code becomes tricky as it is hard to spot an issue if the users implement only one of these functions.



> Eventually, the idea is to remove all children of CostModelBase.

I have already solved this issue. But my plan is to create a follow-up PR with this deprecation code (the idea is to do not make this PR more complex).

Should you agree to mark this conversation as resolved?

Reply:
> I think the way forward would be to do the class hierarchy CostModelAbstract -> CostModelResidual -> CostModelCom..., ... ,

This could be an approach to do. However, I would prefer to avoid this task. Note that each `CostModelCom,...` class will be deprecated. I have already developed this code, but I don't want to push here as it modifies many more files.



> I am okay if you push your implementation in this PR. It would provide a more coherent idea of what you are proposing.

Thank you for the availability (despite the dimension of this PR). But I really prefer to do it in another PR, so you can easily spot any issue. (I want to have this PR merged asap)


I believe I can mark this discussion as resolved. Please feel free to re-open it if you do not believe so!

Reply:
In that case, let's remove this costmodel from this PR. You could add it in the next PR, which would give a better understanding of the final structure.

Reply:
> In that case, let's remove this costmodel from this PR

What do you mean with *costmodel*? Do you mean `CostModelResidual`? (it is the only cost model that has been included in this PR since you've requested in first revision)

I know it is very stressful to review a large PR; however, removing code that then I will include in another PR it is an unnecessary work. If there is something that concerns you, then I am happy to clarify these bits in the final structure ;)

Note that I will create a follow-up PR (just immediately after merging this PR), with the deprecation code. Indeed, this PR is not changing any structure, instead it is adding a new `CostModelResidual` option.

I believe this discussion is resolved. But I am very keen to answer any question that you might still have about the structure.